clients/cli/src/workers/online.rs

```diff
@@ -153,74 +206,111 @@ async fn attempt_task_fetch(
             crate::events::EventType::Refresh,
             LogLevel::Info,
         ))
         .await;
 
-    // Add timeout to prevent hanging
-    let fetch_future = fetch_task_batch(
-        orchestrator_client,
-        node_id,
-        verifying_key,
-        BATCH_SIZE,
-        event_sender,
-    );
-    let timeout_duration = Duration::from_secs(60); // 60 second timeout
+    // Fetch one task with timeout
+    let timeout_duration = Duration::from_secs(60);
+    let node_id_str = node_id.to_string();
+    let fetch_future = orchestrator_client.get_proof_task(&node_id_str, verifying_key);
 
     match tokio::time::timeout(timeout_duration, fetch_future).await {
-        Ok(fetch_result) => match fetch_result {
-            Ok(tasks) => {
-                // Record successful fetch attempt timing
-                state.record_fetch_attempt();
-                handle_fetch_success(
-                    tasks,
-                    sender,
+        Ok(Ok(task)) => {
+            // Check for duplicate
+            if recent_tasks.contains(&task.task_id).await {
+                state.increase_backoff_for_error();
+                send_event(
                     event_sender,
-                    recent_tasks,
-                    state,
-                    environment,
-                    client_id,
+                    format!(
+                        "Task was duplicate - backing off for {}s",
+                        state.backoff_duration().as_secs()
+                    ),
+                    crate::events::EventType::Refresh,
+                    LogLevel::Warn,
                 )
-                .await
+                .await;
+                return Ok(());
             }
-            Err(e) => {
-                // Record failed fetch attempt timing
-                state.record_fetch_attempt();
-                handle_fetch_error(e, event_sender, state).await;
-                Ok(())
+
+            // Add to cache and queue
+            recent_tasks.insert(task.task_id.clone()).await;
+
+            if sender.send(task.clone()).await.is_err() {
+                let _ = event_sender
+                    .send(Event::task_fetcher(
+                        "Task queue is closed".to_string(),
+                        crate::events::EventType::Shutdown,
+                    ))
+                    .await;
+                return Err(true); // Signal shutdown
             }
-        },
-        Err(_timeout) => {
-            // Handle timeout case
-            state.record_fetch_attempt();
+
+            // Track analytics
+            tokio::spawn(track_got_task(
+                task,
+                environment.clone(),
+                client_id.to_string(),
+            ));
+
+            // Success: reset backoff and log
+            state.reset_backoff();
+            let current_queue_level = TASK_QUEUE_SIZE - sender.capacity();
```

Comment:
Overall I think having helper functions for a lot of this stuff was a better overall design approach than trying to squeeze it into a big function like this.